File: src/spice2x/api/resources/python/spiceapi/lights.py
from .connection import Connection
from .request import Request


def lights_read(con: Connection, light_names=None):
    req = Request("lights", "read")

    if light_names:
        for light_name in light_names:
            req.add_param(light_name)

    res = con.request(req)
    return res.get_data()


def lights_write(con: Connection, light_state_list):
    req = Request("lights", "write")
    for state in light_state_list:
        req.add_param(state)
    con.request(req)


def lights_write_reset(con: Connection, light_names=None):
    req = Request("lights", "write_reset")

    # reset all lights
    if not light_names:
        con.request(req)
        return

    # reset specified lights
    for light_name in light_names:
        req.add_param(light_name)
    con.request(req)

File: src/spice2x/api/resources/dart/spiceapi-websocket/src/wrappers/lights.dart
part of spiceapi;

class LightState {
  String name;
  double state;
  bool active;

  LightState(this.name, this.state);
  LightState._fromRead(this.name, this.state, this.active);
}

Future<List<LightState>> lightsRead(Connection con) {
  var req = Request("lights", "read");
  return con.request(req).then((res) {

    // build states list
    List<LightState> states = [];
    for (List state in res.getData()) {
      states.add(LightState._fromRead(
        state[0],
        state[1],
        state[2],
      ));
    }

    // return it
    return states;
  });
}

Future<void> lightsWrite(Connection con, List<LightState> states) {
  var req = Request("lights", "write");

  // add params
  for (var state in states) {
    var obj = [
      state.name,
      state.state
    ];
    req.addParam(obj);
  }

  return con.request(req);
}

Future<void> lightsWriteReset(Connection con, List<String> names) {
  var req = Request("lights", "write_reset");

  // add params
  for (var name in names)
    req.addParam(name);

  return con.request(req);
}
